galley/pkg/server/components/monitoring.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package components
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+
+	ocprom "contrib.go.opencensus.io/exporter/prometheus"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opencensus.io/stats/view"
+
+	"istio.io/pkg/version"
+
+	"istio.io/istio/galley/pkg/server/process"
+)
+
+const (
+	metricsPath = "/metrics"
+	versionPath = "/version"
+)
+
+// NewMonitoring returns a new monitoring component.
+func NewMonitoring(port uint) process.Component {
+
+	var lis net.Listener
+	var exporter *ocprom.Exporter
+	var server *http.Server
+
+	return process.ComponentFromFns(
+		// start
+		func() error {
+			var err error
+			if lis, err = netListen("tcp", fmt.Sprintf(":%v", port)); err != nil {
+				err = fmt.Errorf("unable to listen on monitoring port %v: %v", port, err)
+				return err
+			}
+
+			registry := prometheus.DefaultRegisterer.(*prometheus.Registry)
+			if exporter, err = ocprom.NewExporter(ocprom.Options{Registry: registry}); err != nil {
+				err = fmt.Errorf("could not set up prometheus exporter: %v", err)
+				return err
+			}
+			view.RegisterExporter(exporter)
+
+			mux := http.NewServeMux()
+			mux.Handle(metricsPath, exporter)
+			mux.HandleFunc(versionPath, func(out http.ResponseWriter, req *http.Request) {
+				if _, err := out.Write([]byte(version.Info.String())); err != nil {
+					scope.Errorf("Unable to write version string: %v", err)
+				}
+			})
+
+			version.Info.RecordComponentBuildTag("galley")
+
+			server = &http.Server{
+				Handler: mux,
+			}
+
+			var wg sync.WaitGroup
+			wg.Add(1)
+
+			go func() {
```

Comment:
There is no guarantee that the go routine will start before the function terminates. In the case of an immediate start/stop, it is possible that the go routine might not have started after the call to stop(). 
WaitGroup ensures that the go-routine has started and captures relevant state, before relinquishing control back to the caller of start.

Reply:
Got it.

Does the "state capture" aspect behave differently if arguments are passed to the goroutine function? Or is there still the same issue?

Reply:
No there is no issue if it is a function call. the issue is due to closure. go routine's closure captures pointers to the local variables here.

if the code is of the form:
```
go fn(foo, bar)
```
Then foo and bar is evaluated eagerly (i.e. copied), before calling into the goroutine.

Reply:
That's what I thought. 

So the remaining reason to use a wait group (instead of passing arguments) is to handle the stop-before-start race condition, right? 

Isn't it still technically possible (although I'm sure extremely unlikely) to get them out of sync if the goroutine happens to pause right after calling wg.Done() but before starting the server? 

As a thought exercise, could the "guarantee start before stop" aspect be handled in ComponentFromFns itself, so all new components get that expected behavior? e.g.

```go
func ComponentFromFns(startFn func() error, stopFn func()) Component {
	started := make(chan bool)
	return &component{
		startFn: func() error {
			err := startFn()
			close(started)
			return err
		},
		stopFn:  func() {
			<-started
			stopFn()
		},
	}
}
```

Thanks again for your patience with questions.

Reply:
If the server is Stop'ped even before .Serve is called, .Serve should return immediately with an error. So semantically, the timings shouldn't matter.